feat(synthetic): Synthetic test environment

flynncs opened the pull request:
### Changes

- New synthetic-tests package
  - Utilises the artillery package to run load tests against a target deployment, simulating 'real' user behaviour
- New fake-data methods
  - Generate fake data for API requests using zod schemas and the zod-mock library
- More schemas in facility-server package
  - Supports easier generation of fake API request data in fake-data
- Fix Playwright compilation / importing errors
  - Playwright now runs in CJS, so it can import from other packages such as fake-data and constants

### Deploys

- [x] **Deploy to Tamanu Internal** <!-- #deploy -->

### Tests

- [x] **Run E2E Tests** <!-- #e2e -->

### Synthetic Test Environment

- [ ] **Synthetic Test Environment** <!-- #synthetic-test -->

**Target URL:** https://facility-1.feature-synthetic-test-setup.cd.tamanu.app

### Remember to...

- ...write or update tests
- ...add UI screenshots and **testing notes** to the Linear issue
- ...add any **manual upgrade steps** to the Linear issue
- ...update the [config reference](https://beyond-essential.slab.com/posts/reference-config-file-0c70ukly), [settings reference](https://beyond-essential.slab.com/posts/reference-settings-0blw1x2q), or any [relevant runbook(s)](https://beyond-essential.slab.com/topics/runbooks-bs04ml6c)
- ...call out additions or changes to **config files** for the deployment team to take note of

<!-- Thank you! -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a comprehensive synthetic testing framework, including new configuration files, reusable GitHub Actions workflows, and automated scenario merging.
  - Added utilities and hooks for generating realistic fake patient and encounter data for tests.
  - Implemented automated validation and load testing for patient and encounter creation endpoints.

- **Bug Fixes**
  - Improved validation and error handling for patient and encounter API endpoints.

- **Documentation**
  - Added detailed procedural guides for creating synthetic test scenarios.

- **Chores**
  - Updated package configurations, dependencies, and workspace settings to support new testing infrastructure.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

flynncs committed:
basic impl

flynncs committed:
ci def

flynncs committed:
Merge branch 'main' into feature/synthetic-test-setup

flynncs committed:
merge

flynncs committed:
test target

flynncs committed:
remove module type

flynncs committed:
global install for artillery @ fix ci output

flynncs committed:
fix gh token

flynncs committed:
should retrigger better now

flynncs committed:
state of working - new schemas & fake data

flynncs committed:
fake foreign key fixes

flynncs committed:
in a state of working now

flynncs committed:
use the generateId util

flynncs committed:
ensure required fields are passed in

flynncs committed:
update actions

flynncs committed:
fix pipeline

flynncs committed:
remove target override

flynncs committed:
fix target arg

flynncs committed:
retry

flynncs committed:
fix url

flynncs committed:
new re-trigger method

flynncs committed:
fix sha

flynncs committed:
encounter scenario

flynncs committed:
merge logic

flynncs committed:
Merge branch 'main' into feature/synthetic-test-setup

flynncs committed:
package

flynncs committed:
Merge branch 'feature/synthetic-test-setup' of https://github.com/beyondessential/tamanu into feature/synthetic-test-setup

flynncs committed:
better fake request utils

flynncs committed:
ci updates

flynncs committed:
fix calls to fake request body

flynncs committed:
Merge branch 'main' into feature/synthetic-test-setup

flynncs committed:
fix fail condition

flynncs committed:
cancel runs for same head branch

flynncs committed:
specify gh repo

flynncs committed:
pr friendly for now

flynncs committed:
fix target regex

flynncs committed:
use native dispatcher

flynncs committed:
actually we might need to do the other way

flynncs committed:
use gh workflow instead

flynncs committed:
fix unknown repo

flynncs committed:
register the runner workflow

flynncs committed:
fix inputs

flynncs committed:
debug it

flynncs committed:
fix run command

flynncs committed:
scenario rule

flynncs committed:
update settings schema paths

flynncs committed:
Merge branch 'main' into feature/synthetic-test-setup

flynncs committed:
fix merge lock

flynncs committed:
fix no patient found

flynncs committed:
add expect plugin to config

flynncs committed:
fix runner retrigger

dannash100 commented on .github/workflows/ci-synthetic-test-runner.yml:
```diff
@@ -0,0 +1,150 @@
+    runs-on: ubuntu-latest
+    env:
+      TARGET_URL: ${{ github.event.inputs.target_url }}
+      PR_NUMBER: ${{ github.event.inputs.pr_number }}
```
Not certain tho I might get Felix to have a look at this side of things

dannash100 commented on .github/workflows/ci-synthetic-test-trigger.yml:
```diff
@@ -0,0 +1,71 @@
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          echo "Triggering ci-synthetic-test-runner.yml on ref ${{ steps.check.outputs.pr_ref }}"
```
You can trigger workflows like this
```
    uses: ./.github/workflows/ci-synthetic-test-runner.yml
    with:
      ....
```

dannash100 commented on .github/workflows/ci-synthetic-test-runner.yml:
```diff
@@ -0,0 +1,150 @@
+    name: synthetic-test-run-${{ github.event.inputs.pr_number }}-${{ github.run_id }}
+    runs-on: ubuntu-latest
+    env:
+      TARGET_URL: ${{ github.event.inputs.target_url }}
```
Im thinking we could default target url to the auto-deploy facility url, if you check out the android cd action it does this.

dannash100 commented on packages/e2e-tests/utils/generateNewPatient.ts:
```diff
@@ -1,38 +1,24 @@
+    },
+  });
+
+  console.log('Patient data:', patientData);
+
```
```suggestion
```

dannash100 commented on packages/facility-server/app/routes/apiv1/encounter.js:
```diff
@@ -47,7 +49,8 @@ encounter.post(
     const { models, body, user } = req;
     req.checkPermission('create', 'Encounter');
-    const encounterObject = await models.Encounter.create({ ...body, actorId: user.id });
+    const validatedBody = validate(createEncounterSchema, body);
```
Note to check if this should be async validation.

dannash100 commented on packages/facility-server/app/routes/apiv1/encounter.schema.js:
```diff
@@ -0,0 +1,37 @@
+
+export const createEncounterSchema = z.object({
+  // Required fields for encounter creation
+  encounterType: z.enum([
```
This could just be Object.keys(ENCOUNTER_TYPES) right

dannash100 commented on packages/facility-server/app/routes/apiv1/encounter.schema.js:
```diff
@@ -0,0 +1,37 @@
+  departmentId: foreignKey,
+
+  // Optional fields
+  endDate: z.coerce.date().optional(), // ISO datetime string
```
I notice that other usages of zod are using `datetimeCustomValidation` see if we can standardize that here.

dannash100 commented on packages/facility-server/app/routes/apiv1/patient/patient.schema.js:
```diff
@@ -0,0 +1,167 @@
+
+export const createPatientSchema = z.object({
+  // Required fields for patient creation
+  patientRegistryType: z.enum([
```
Can use obejct values for these enums

dannash100 commented on packages/facility-server/app/routes/apiv1/patient/patient.schema.js:
```diff
@@ -0,0 +1,167 @@
+  birthType: z.enum([BIRTH_TYPES.SINGLE, BIRTH_TYPES.PLURAL]).optional(),
+  birthWeight: z.number().optional(),
+  birthLength: z.number().optional(),
+  apgarScoreOneMinute: z.number().int().min(0).max(10).optional(),
```
I think these ones are all min 1 actully

dannash100 commented on packages/facility-server/app/routes/apiv1/patient/patient.schema.js:
```diff
@@ -0,0 +1,167 @@
+
+  // Birth Details (from GenericBirthFields - only when patientRegistryType is 'birth_registry')
+  timeOfBirth: z.coerce.date().optional(), // ISO datetime string
+  gestationalAgeEstimate: z.number().optional(),
```
This ones is
      .min(1)
        .max(45) on web btw 
      
And birth length and a few others are also min maxed on web side.
If we added these here lets do all I rekon

dannash100 reviewed (changes requested):
Just leaving my review in process here

passcod commented on .github/workflows/ci-synthetic-test-runner.yml:
```diff
@@ -0,0 +1,150 @@
+        uses: actions/github-script@v7
+        with:
+          script: |
+            console.log('PR_NUMBER from env:', process.env.PR_NUMBER);
```
GHA has a debug log thing too, if you want to have things clean by default but still present long term if we ever need. That said, env variables are directly available in the log output regardless of you printing them, so this isn't really ever needed.

passcod commented on .github/workflows/ci-synthetic-test-runner.yml:
```diff
@@ -0,0 +1,150 @@
+              target_url: `https://github.com/${context.repo.owner}/${context.repo.repo}/actions/runs/${process.env.GITHUB_RUN_ID}`
+            });
+
+      - name: Retrigger synthetic test 
```
(Noting that we're only burning microsoft's dollars here, but it's still good to consider in a good-citizen sense)

edmofro replied:
Plus there's an acceptable use clause, they could crack down on us if we push the boat too far

dannash100 replied:
ohh omg I forgot we don't pay for actions any more!

passcod commented on .github/workflows/ci-synthetic-test-trigger.yml:
```diff
@@ -0,0 +1,71 @@
+              pull_number: prNumber
+            });
+
+            const syntheticTestMatch = pr.body?.match(/\[(?<checked>[ x])\]\s+\*\*Synthetic Test Environment\*\* <!-- #synthetic-test -->/);
```
I also would make it so that it matches the comment and the checkbox, but allows the visible text to be anything

flynncs committed:
remove logs

dannash100 reviewed (commented):
Few more

dannash100 reviewed (changes requested):
Last few bits of review! Also check out the code rabbit ones, some look interesting. Amazing work yeye

flynncs committed:
call & max attempts

flynncs committed:
better should run matching

flynncs committed:
schema updates

flynncs committed:
fix trigger ref

flynncs committed:
re fix the ref

flynncs committed:
fix cancel logic

flynncs committed:
add back runs on

flynncs committed:
swap to gh script

flynncs committed:
revert back to dispatch

flynncs committed:
fix missing ref

flynncs committed:
fix runner

flynncs committed:
fix cancellation logic

flynncs committed:
fix location and department id missing

flynncs committed:
some suggestions

flynncs committed:
coderabbit

flynncs committed:
Merge branch 'main' into feature/synthetic-test-setup

flynncs committed:
Merge branch 'main' into feature/synthetic-test-setup

flynncs committed:
Merge branch 'main' into feature/synthetic-test-setup

flynncs committed:
Migrate to zod v4 and centralize request schemas

Upgrades all packages to use zod v4, replacing deprecated zod-mock with zocker for mock data generation. Moves encounter and patient request schemas to the shared package for better reuse and updates imports accordingly. Refactors schema types for improved validation, adds max length constraints, and enhances validation error logging. Updates dependencies and removes unused packages.

flynncs committed:
Merge branch 'main' into feature/synthetic-test-setup

flynncs committed:
Update package-lock.json

dannash100 reviewed (approved):
!! Gotem thank u

flynncs committed:
remove expect & add metrics by endpoint

flynncs committed:
Merge branch 'main' into feature/synthetic-test-setup

flynncs committed:
fix retrigger?

flynncs committed:
Merge branch 'main' into feature/synthetic-test-setup

flynncs committed:
remove template

flynncs committed:
Merge branch 'feature/synthetic-test-setup' of https://github.com/beyondessential/tamanu into feature/synthetic-test-setup

flynncs committed:
console log

flynncs committed:
Merge branch 'main' into feature/synthetic-test-setup

flynncs committed:
soft parse import ongoing medications

flynncs committed:
Merge branch 'main' into feature/synthetic-test-setup

flynncs committed:
Merge branch 'main' into feature/synthetic-test-setup

flynncs committed:
revert e2e generators for now
